File: services/ai/utilities/self_consistency.ts
import { complexity_score } from '../pipeline';
import { validateAgainstSchema, SchemaType } from '../schemas/validator';

export interface SelfConsistencyConfig {
  enabled: boolean;
  k: number; // number of samples
  temperatures: number[]; // varied temperatures
  minComplexity: number;
  targetSchema: SchemaType;
}

export interface ConsistencyResult<T> {
  majorityVote: T;
  agreementScore: number; // 0-1, higher means more agreement
  samples: T[];
  confidence: number;
}

// Mock LLM call function - would be replaced with actual OpenAI/etc.
async function mockLLMCall(prompt: string, temperature: number, schema?: SchemaType): Promise<any> {
  // Placeholder - would make actual API call
  console.log(`[SelfConsistency] Mock LLM call with temp ${temperature}, schema ${schema}`);
  
  // Return mock structured response
  if (schema === 'plan') {
    return {
      intent: 'plan',
      goal: 'Achieve objective',
      steps: ['Step 1', 'Step 2', 'Step 3'],
      timeline: '1 week',
      resources_needed: ['Resource A', 'Resource B'],
    };
  }
  
  return { mock: true };
}

export class SelfConsistencyVoter {
  private config: SelfConsistencyConfig;

  constructor(config: Partial<SelfConsistencyConfig> = {}) {
    this.config = {
      enabled: config.enabled ?? true,
      k: config.k ?? 5,
      temperatures: config.temperatures ?? [0.3, 0.5, 0.7, 0.9, 1.1],
      minComplexity: config.minComplexity ?? 0.6,
      targetSchema: config.targetSchema ?? 'plan',
    };
  }

  async shouldUseSelfConsistency(input: { userMessage: string; context?: any }, schema: SchemaType): Promise<boolean> {
    if (!this.config.enabled) return false;
    if (schema !== this.config.targetSchema) return false;
    
    const complexity = complexity_score(input);
    return complexity >= this.config.minComplexity;
  }

  async generateConsistentResponse<T>(
    prompt: string,
    schema: SchemaType,
    input: { userMessage: string; context?: any }
  ): Promise<ConsistencyResult<T> | null> {
    const shouldUse = await this.shouldUseSelfConsistency(input, schema);
    if (!shouldUse) return null;

    // Generate k samples with varied temperatures
    const samples: T[] = [];
    if (this.config.temperatures.length === 0) {
      throw new Error('No temperatures configured for self-consistency voting');
    }

    for (let i = 0; i < this.config.k; i++) {
      const index = i % this.config.temperatures.length;
      const temp = this.config.temperatures[index]!;
      try {
        const sample = await mockLLMCall(prompt, temp, schema);
        // Validate sample
        const validation = validateAgainstSchema(schema, sample);
        if (validation.valid) {
          samples.push(sample as T);
        }
      } catch (error) {
        console.warn(`[SelfConsistency] Sample ${i} failed:`, error);
      }
    }

    if (samples.length === 0) {
      throw new Error('No valid samples generated for self-consistency voting');
    }

    // Find majority vote based on structured field agreement
    const majorityVote = this.findMajorityVote(samples);
    const agreementScore = this.calculateAgreementScore(samples, majorityVote);

    return {
      majorityVote,
      agreementScore,
      samples,
      confidence: agreementScore * (samples.length / this.config.k),
    };
  }

  private findMajorityVote<T>(samples: T[]): T {
    if (samples.length === 0) {
      throw new Error('Cannot determine majority vote from empty samples');
    }
    // Simple majority voting - for complex objects, this would need more sophisticated logic
    // For now, return the first sample (placeholder)
    return samples[0]!;
  }

  private calculateAgreementScore<T>(samples: T[], majorityVote: T): number {
    // Calculate how many samples agree with majority vote
    let agreements = 0;
    for (const sample of samples) {
      if (this.samplesAgree(sample, majorityVote)) {
        agreements++;
      }
    }
    return agreements / samples.length;
  }

  private samplesAgree<T>(a: T, b: T): boolean {
    // Simple equality check - for structured data, would compare key fields
    return JSON.stringify(a) === JSON.stringify(b);
  }
}

// Singleton instance
export const selfConsistencyVoter = new SelfConsistencyVoter();

File: services/python-ai/app/ai_legacy/utilities/self_consistency.py
from typing import Any, Dict, List, Optional

from ..pipeline import complexity_score
from ..schemas.validator import SchemaType, validate_against_schema


class SelfConsistencyVoter:
    def __init__(self, enabled: bool = True, k: int = 5, temperatures: Optional[List[float]] = None, min_complexity: float = 0.6, target_schema: SchemaType = "plan") -> None:
        self.config = {
            "enabled": enabled,
            "k": k,
            "temperatures": temperatures or [0.3, 0.5, 0.7, 0.9, 1.1],
            "minComplexity": min_complexity,
            "targetSchema": target_schema,
        }

    async def should_use_self_consistency(self, input_payload: Dict[str, Any], schema: SchemaType) -> bool:
        if not self.config["enabled"]:
            return False
        if schema != self.config["targetSchema"]:
            return False
        complexity = complexity_score(input_payload)
        return complexity >= self.config["minComplexity"]

    async def generate_consistent_response(self, prompt: str, schema: SchemaType, input_payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        if not await self.should_use_self_consistency(input_payload, schema):
            return None
        samples: List[Any] = []
        temps = self.config["temperatures"]
        if not temps:
            raise ValueError("No temperatures configured for self-consistency voting")
        for i in range(self.config["k"]):
            temp = temps[i % len(temps)]
            sample = await self._mock_llm_call(prompt, temp, schema)
            validation = validate_against_schema(schema, sample)
            if validation["valid"]:
                samples.append(sample)
        if not samples:
            raise ValueError("No valid samples generated for self-consistency voting")
        majority_vote = samples[0]
        agreement_score = self._calculate_agreement_score(samples, majority_vote)
        return {
            "majorityVote": majority_vote,
            "agreementScore": agreement_score,
            "samples": samples,
            "confidence": agreement_score * (len(samples) / self.config["k"]),
        }

    async def _mock_llm_call(self, prompt: str, temperature: float, schema: SchemaType) -> Any:
        if schema == "plan":
            return {
                "intent": "plan",
                "goal": "Achieve objective",
                "steps": ["Step 1", "Step 2", "Step 3"],
                "timeline": "1 week",
                "resources_needed": ["Resource A", "Resource B"],
            }
        return {"mock": True}

    def _calculate_agreement_score(self, samples: List[Any], majority_vote: Any) -> float:
        agreements = 0
        for sample in samples:
            if self._samples_agree(sample, majority_vote):
                agreements += 1
        return agreements / len(samples)

    def _samples_agree(self, a: Any, b: Any) -> bool:
        return a == b


self_consistency_voter = SelfConsistencyVoter()
